src/modules/bulletproof/rangeproof_impl.h

```diff
@@ -0,0 +1,547 @@
+        secp256k1_ecmult_const(&stermj, &genh[i], &sr, 256);
+        secp256k1_ge_set_gej(&sterm, &stermj);
+        secp256k1_gej_add_ge(&sj, &sj, &sterm);
+    }
+    secp256k1_rfc6979_hmac_sha256_finalize(&lr_gen.rng);
+
+    /* get challenges y and z */
+    secp256k1_ge_set_gej(&out_pt[0], &aj);
+    secp256k1_ge_set_gej(&out_pt[1], &sj);
+
+    secp256k1_bulletproof_update_commit(commit, &out_pt[0], &out_pt[1]);
+    secp256k1_scalar_set_b32(&y, commit, &overflow);
+    if (overflow || secp256k1_scalar_is_zero(&y)) {
+        return 0;
+    }
+    secp256k1_bulletproof_update_commit(commit, &out_pt[0], &out_pt[1]); /* TODO rehashing A and S to get a second challenge is overkill */
+    secp256k1_scalar_set_b32(&z, commit, &overflow);
+    if (overflow || secp256k1_scalar_is_zero(&z)) {
+        return 0;
+    }
+    secp256k1_scalar_sqr(&zsq, &z);
+
+    /* Compute coefficients t0, t1, t2 of the <l, r> polynomial */
+    /* t0 = l(0) dot r(0) */
+    secp256k1_lr_generator_init(&lr_gen, &rng, &y, &z, value);
+    secp256k1_scalar_clear(&t0);
+    for (i = 0; i < nbits; i++) {
+        secp256k1_scalar l, r;
+        secp256k1_lr_generate(&lr_gen, &l, &r, &zero);
+        secp256k1_scalar_mul(&l, &l, &r);
+        secp256k1_scalar_add(&t0, &t0, &l);
+    }
+    secp256k1_lr_generator_finalize(&lr_gen);
+
+    /* A = t0 + t1 + t2 = l(1) dot r(1) */
+    secp256k1_lr_generator_init(&lr_gen, &rng, &y, &z, value);
+    secp256k1_scalar_clear(&t1);
+    for (i = 0; i < nbits; i++) {
+        secp256k1_scalar one;
+        secp256k1_scalar l, r;
+        secp256k1_scalar_set_int(&one, 1);
+        secp256k1_lr_generate(&lr_gen, &l, &r, &one);
+        secp256k1_scalar_mul(&l, &l, &r);
+        secp256k1_scalar_add(&t1, &t1, &l);
+    }
+    secp256k1_lr_generator_finalize(&lr_gen);
+
+    /* B = t0 - t1 + t2 = l(-1) dot r(-1) */
+    secp256k1_lr_generator_init(&lr_gen, &rng, &y, &z, value);
+    secp256k1_scalar_clear(&t2);
+    for (i = 0; i < nbits; i++) {
+        secp256k1_scalar negone;
+        secp256k1_scalar l, r;
+        secp256k1_scalar_set_int(&negone, 1);
+        secp256k1_scalar_negate(&negone, &negone);
+        secp256k1_lr_generate(&lr_gen, &l, &r, &negone);
+        secp256k1_scalar_mul(&l, &l, &r);
+        secp256k1_scalar_add(&t2, &t2, &l);
+    }
+    secp256k1_lr_generator_finalize(&lr_gen);
+
+    /* t1 = (A - B)/2 */
+    secp256k1_scalar_set_int(&tmps, 2);
+    secp256k1_scalar_inverse_var(&tmps, &tmps);
+    secp256k1_scalar_negate(&t2, &t2);
+    secp256k1_scalar_add(&t1, &t1, &t2);
+    secp256k1_scalar_mul(&t1, &t1, &tmps);
+
+    /* t2 = -(-B + t0) + t1 */
+    secp256k1_scalar_add(&t2, &t2, &t0);
+    secp256k1_scalar_negate(&t2, &t2);
+    secp256k1_scalar_add(&t2, &t2, &t1);
+
+    /* Compute Ti = t_i*A + tau_i*G for i = 1,2 */
+    secp256k1_gej_set_ge(&tmpj, genp);
+    secp256k1_ecmult(ecmult_ctx, &tj[0], &tmpj, &t1, &tau1);
+    secp256k1_ecmult(ecmult_ctx, &tj[1], &tmpj, &t2, &tau2);
```

Comment:
Both these `ecmult` need to be constant time since `tau1` and `tau2` are secrets.